### PR TITLE
TMDM-12841 Don't allow redundant value for operator "Is Empty Or Null" in view's where condition.

### PR DESCRIPTION
**Pull request by sbliu**
https://jira.talendforge.org/browse/TMDM-12841

**What is the current behavior?** (You should also link to an open issue here)
when view condition operator is "Is Empty Or Null", value is allow as other operators

**What is the new behavior?**
 when view condition operator is "Is Empty Or Null", value is not allowed.

**Please check if the PR fulfills these requirements**

- [x] The commit message follows Talend standard
- [ ] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features) ?

**What kind of change does this PR introduce?**

- [x] Bugfix
- [ ] Feature
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build / CI related changes
- [ ] Other... Please describe:

**Does this PR introduce a breaking change?**

- [ ] Yes
- [x] No

If this PR contains a breaking change, please describe the impact and migration path for existing applications: ...

**Other information**:

### CODE REVIEW

File: main/plugins/org.talend.mdm.repository.validate/src/org/talend/mdm/repository/core/validate/i18n/messages.properties [hunk 1]
```diff
@@ -22,4 +22,5 @@ ValidationResultDialog_NotShowThis=Don't show this dialog again.
 ValidationResultDialog_skipAllErrs=Skip All Errors
 ValidationResultDialog_skipAllErrWarnings=Skip All Errors/Warnings
 ValidationResultDialog_Title=Validation Result Dialog
+ViewValidator_EmptyNullOperatorValue_error=Value in where condition of view "{0}" is unecessary for operator on XPath "{1}".
```
Comment (lpw82gf) [inline main/plugins/org.talend.mdm.repository.validate/src/org/talend/mdm/repository/core/validate/i18n/messages.properties]: `unecessary` to `unnecessary` ?

File: main/plugins/org.talend.mdm.repository/src/main/java/org/talend/mdm/repository/ui/widgets/UserVarValueValidator.java [hunk 2]
```diff
@@ -3,36 +3,34 @@
...
+
+    private static final String _PREFIX_USER_VAR = "${user_context."; //$NON-NLS-1$
+
+    private static final String _SURFIX_USER_VAR = "}"; //$NON-NLS-1$
+
+    /**
+     * True: Is valid
+     */
+    public static boolean validate(String userVarValue) {
+        List<String> validUserVars = new ArrayList<String>();
```
Comment (lpw82gf) [inline main/plugins/org.talend.mdm.repository/src/main/java/org/talend/mdm/repository/ui/widgets/UserVarValueValidator.java]: Change to `List<String> validUserVars = new ArrayList<>();`

File: main/plugins/org.talend.mdm.repository.validate/src/org/talend/mdm/repository/core/validate/i18n/messages.properties (hunk 1 above)
Comment (sbliu) [reply main/plugins/org.talend.mdm.repository.validate/src/org/talend/mdm/repository/core/validate/i18n/messages.properties]: yes

File: main/plugins/org.talend.mdm.repository/src/main/java/org/talend/mdm/repository/ui/widgets/UserVarValueValidator.java (hunk 2 above)
Comment (sbliu) [reply main/plugins/org.talend.mdm.repository/src/main/java/org/talend/mdm/repository/ui/widgets/UserVarValueValidator.java]: We think this change is not necessary.

Comment (lpw82gf) [reply main/plugins/org.talend.mdm.repository/src/main/java/org/talend/mdm/repository/ui/widgets/UserVarValueValidator.java]: We have switch to Java 8 for a long time, and we should use new style to define a List. Please.

File: main/plugins/org.talend.mdm.repository/src/main/java/org/talend/mdm/repository/ui/widgets/UserVarValueValidator.java [hunk 3]
```diff
@@ -3,36 +3,34 @@
...
+    public static boolean validate(String userVarValue) {
+        List<String> validUserVars = new ArrayList<>();
+        validUserVars.add(_PREFIX_USER_VAR + UserField.Id.field + _SURFIX_USER_VAR);
+        validUserVars.add(_PREFIX_USER_VAR + UserField.First_Name.field + _SURFIX_USER_VAR);
+        validUserVars.add(_PREFIX_USER_VAR + UserField.Last_Name.field + _SURFIX_USER_VAR);
+        validUserVars.add(_PREFIX_USER_VAR + UserField.User_Name.field + _SURFIX_USER_VAR);
+        validUserVars.add(_PREFIX_USER_VAR + UserField.Language.field + _SURFIX_USER_VAR);
+
+        boolean result = true;
+        if (userVarValue.startsWith(_PREFIX_USER_VAR) && userVarValue.endsWith(_SURFIX_USER_VAR)) {
```
Comment (YohannTalend) [inline main/plugins/org.talend.mdm.repository/src/main/java/org/talend/mdm/repository/ui/widgets/UserVarValueValidator.java]: Why not to concatenate the three conditions ?

File: main/plugins/org.talend.mdm.repository.validate/src/org/talend/mdm/repository/core/validate/i18n/messages.properties [hunk 4]
```diff
@@ -22,4 +22,5 @@ ValidationResultDialog_NotShowThis=Don't show this dialog again.
 ValidationResultDialog_skipAllErrs=Skip All Errors
 ValidationResultDialog_skipAllErrWarnings=Skip All Errors/Warnings
 ValidationResultDialog_Title=Validation Result Dialog
+ViewValidator_EmptyNullOperatorValue_error=Value in where condition of view "{0}" is unnecessary for operator on XPath "{1}".
```
Comment (pbabin) [inline main/plugins/org.talend.mdm.repository.validate/src/org/talend/mdm/repository/core/validate/i18n/messages.properties]: If it is an error, isn't it the message below instead?
`Error on view "{0}" on XPath "{1}": using "Is Empty Or Null" operator, it is not allowed to set a condition value.`
Looks like message you give just says it's not necessary but isn't harmful.

**Comment by pbabin** [comment]
BTW I don't think we are fixing the real issue (which looks like minor) here. It seems somehow editor can accept an empty String as a value which shouldn't be the case.

File: main/plugins/org.talend.mdm.repository/src/main/java/org/talend/mdm/repository/ui/widgets/UserVarValueValidator.java (hunk 3 above)
Comment (sbliu) [reply main/plugins/org.talend.mdm.repository/src/main/java/org/talend/mdm/repository/ui/widgets/UserVarValueValidator.java]: "three conditions" means?

Comment (VictorLi-talendbj) [reply main/plugins/org.talend.mdm.repository/src/main/java/org/talend/mdm/repository/ui/widgets/UserVarValueValidator.java]: I think it means we can try to combine the three 'if' conditions into a single one.

File: main/plugins/org.talend.mdm.repository.validate/src/org/talend/mdm/repository/core/validate/i18n/messages.properties (hunk 4 above)
Comment (sbliu) [reply main/plugins/org.talend.mdm.repository.validate/src/org/talend/mdm/repository/core/validate/i18n/messages.properties]: Corrected as you said.

File: main/plugins/org.talend.mdm.repository/src/main/java/org/talend/mdm/repository/ui/widgets/UserVarValueValidator.java [hunk 5]
```diff
@@ -3,36 +3,33 @@
...
+        validUserVars.add(_PREFIX_USER_VAR + UserField.User_Name.field + _SURFIX_USER_VAR);
+        validUserVars.add(_PREFIX_USER_VAR + UserField.Language.field + _SURFIX_USER_VAR);
+
+        String propertyFieldHead = _PREFIX_USER_VAR + UserField.Properties.field + "[\""; //$NON-NLS-1$
+        String propertyFieldTail = "\"]" + _SURFIX_USER_VAR; //$NON-NLS-1$
+
+        boolean isUserVarPattern = userVarValue.startsWith(_PREFIX_USER_VAR) && userVarValue.endsWith(_SURFIX_USER_VAR);
+        boolean propPattern = userVarValue.startsWith(propertyFieldHead) && userVarValue.endsWith(propertyFieldTail)
+                && userVarValue.length() > propertyFieldHead.length() + propertyFieldTail.length();
+        if (isUserVarPattern && !validUserVars.contains(userVarValue) && !propPattern) {
```
[the remark this answers is not in the document]
Comment (sbliu) [reply main/plugins/org.talend.mdm.repository/src/main/java/org/talend/mdm/repository/ui/widgets/UserVarValueValidator.java]: As YohannTalend said, to concatenate the three conditions

File: main/plugins/org.talend.mdm.repository/src/main/java/org/talend/mdm/repository/ui/widgets/UserVarValueValidator.java [hunk 6]
```diff
@@ -3,36 +3,33 @@
...
-                        result = false;
-                    }
-                }
-            }
-
-            return result;
+
+    private static final String _PREFIX_USER_VAR = "${user_context."; //$NON-NLS-1$
+
+    private static final String _SURFIX_USER_VAR = "}"; //$NON-NLS-1$
```
[the remark this answers is not in the document]
Comment (sbliu) [reply main/plugins/org.talend.mdm.repository/src/main/java/org/talend/mdm/repository/ui/widgets/UserVarValueValidator.java]: done

**Pull request merged**